bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/action/EcobeeActions.java

```diff
@@ -0,0 +1,783 @@
+            return false;
+        }
+    }
+
+    /**
+     * Set a hold
+     */
+    @RuleAction(label = "Set Hold", description = "The set hold function sets the thermostat into a hold with the specified temperature or climate ref.")
+    public @ActionOutput(name = "success", type = "java.lang.Boolean") Boolean setHold(
+            @ActionInput(name = "coolHoldTemp", description = "(opt) The temperature at which to set the cool hold.") @Nullable QuantityType<Temperature> coolHoldTemp,
+            @ActionInput(name = "heatHoldTemp", description = "(opt) The temperature at which to set the heat hold.") @Nullable QuantityType<Temperature> heatHoldTemp,
+            @ActionInput(name = "holdClimateRef", description = "(opt) The Climate to use as reference for setting the coolHoldTemp, heatHoldTemp and fan settings for this hold. If this value is passed the coolHoldTemp and heatHoldTemp are not required.") @Nullable String holdClimateRef,
+            @ActionInput(name = "startDateTime", description = "(opt) The start date in thermostat time.") @Nullable Date startDateTime,
+            @ActionInput(name = "endDateTime", description = "(opt) The end date in thermostat time.") @Nullable Date endDateTime,
+            @ActionInput(name = "holdType", description = "(opt) The hold duration type. Valid values: dateTime, nextTransition, indefinite, holdHours.") @Nullable String holdType,
+            @ActionInput(name = "holdHours", description = "(opt) The number of hours to hold for, used and required if holdType='holdHours'.") @Nullable Number holdHours) {
+        Map<String, Object> params = new HashMap<String, Object>();
+        if (coolHoldTemp != null) {
+            params.put("coolHoldTemp", coolHoldTemp);
+        }
+        if (heatHoldTemp != null) {
+            params.put("heatHoldTemp", heatHoldTemp);
+        }
+        if (holdClimateRef != null) {
+            params.put("holdClimateRef", holdClimateRef);
+        }
+        return setHold(params, holdType, holdHours, startDateTime, endDateTime);
+    }
+
+    public static boolean setHold(@Nullable ThingActions actions, @Nullable QuantityType<Temperature> coolHoldTemp,
+            @Nullable QuantityType<Temperature> heatHoldTemp, @Nullable String holdClimateRef,
+            @Nullable Date startDateTime, @Nullable Date endDateTime, @Nullable String holdType,
+            @Nullable Number holdHours) {
+        if (actions == null) {
+            throw new IllegalArgumentException("actions cannot be null");
+        }
+        try {
+            return (Boolean) actions.getClass()
+                    .getDeclaredMethod("setHold", QuantityType.class, QuantityType.class, String.class, Date.class,
+                            Date.class, String.class, Number.class)
+                    .invoke(actions, coolHoldTemp, heatHoldTemp, holdClimateRef, startDateTime, endDateTime, holdType,
+                            holdHours);
```

Comment:
Why is this being done through reflection rather than just casting to a known action class?

Reply:
I did this as a short term workaround for [this issue](https://github.com/openhab/openhab-core/issues/1265). I got tired of restarting openHAB every time I dropped a new version of the binding in addons. As I'm using the binding to control the thermostats in my home, the last thing I want is for the actions to stop working. At some point, when the problem is fixed in core, I'll revert this back to the normal way of doing it.

Reply:
Well that issue is certainly interesting! I did a little online snooping on seeing if we can abuse generics to get around this issue. Something like this might work (using sendMessage as an example here):
```
public static boolean <T extends EcobeeActions> sendMessage(@Nullable ThingActions actions, @Nullable String text) {
        if (actions == null) {
            throw new IllegalArgumentException("actions cannot be null");
        }
        try {
            return ((T) actions).sendMessage(text);
        } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
                | InvocationTargetException e) {
            LOGGER.info("EcobeeActions: Exception getting or invoking method", e);
            return false;
        }
    }
```
The idea is that the compiler can't perform casting at runtime because it has no idea what to cast it to as a generic, so the only thing it can try to do is call that particular method on the object regardless of what type it is.

I suspect this might work but then again the runtime might still try to cast it to upper limit generic value of EcobeeActions anyway. If it actually attempts to cast it this would blow up in our face. 

WDYT?

Reply:
Give me sec to experiment before you try it out, I don't want you to waste your time on something I didn't try first.

Reply:
Thanks for trying another possible solution. I wish there was a less ugly way to deal with this.

Reply:
Well I found a less ugly way to deal with it but it requires the overhead of making an interface.
```

    private static @Nullable IEcobeeActions tryCast(ThingActions actions){
        if(actions.getClass().getName().equals(EcobeeActions.class.getName())){
            if(actions instanceof IEcobeeActions){
                return (IEcobeeActions) actions;
            }else{
                return (IEcobeeActions) Proxy.newProxyInstance(IEcobeeActions.class.getClassLoader(), new Class[]{IEcobeeActions.class}, 
                        (Object proxy, Method method, Object[] args) -> {
                            Method m = actions.getClass().getDeclaredMethod(method.getName(), method.getParameterTypes());
                            return m.invoke(actions, args);
                        });
            }
        }
        return null;
    }
```
This solution is pretty much the same as what you are already doing it just offloads all of the reflection logic into a java proxy. Since java's built-in proxy api only allows for proxying interfaces, you would need to create an interface (IEcobeeActions in this example) with all the action methods.
It would be possible to do away with the interface if we used a proxy library like cglib which can actually handle proxying concrete classes.

This is hardly a proper solution but definitely make the workaround look cleaner. Still don't know if it worth the overhead of creating an interface, but if we really wanted to we could use cglib to proxy the concrete class directly. But importing a library for the sole purpose of a cleaner workaround is a bit iffy. 
WDYT?

Reply:
@cpmeister Another thing I tried was placing all the reflection stuff in a separate static method. This seems to work for me, since all my actions have very simple returns (Boolean, String, etc).

```
    private static <T> Object invokeMethod(ThingActions actions, String name, Class<?>[] classes, Object... args) {
        if (args.length != classes.length) {
            throw new IllegalArgumentException("Number of args and number of classes don't match");
        }
        try {
            Method m = actions.getClass().getDeclaredMethod(name, classes);
            Class<?> returnClass = m.getReturnType();
            return returnClass.cast(m.invoke(actions, args));
        } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
                | InvocationTargetException e) {
            throw new UnsupportedOperationException("Problem getting or invoking method " + name);
        }
    }
```

Then the static action methods would invoke it like this.
```
    public static boolean acknowledge(@Nullable ThingActions actions, @Nullable String ackRef, @Nullable String ackType,
            @Nullable Boolean remindMeLater) {
        if (actions == null) {
            throw new IllegalArgumentException("actions cannot be null");
        }
        Class<?>[] argTypes = { String.class, String.class, Boolean.class };
        return (Boolean) invokeMethod(actions, "acknowledge", argTypes, ackRef, ackType, remindMeLater);
    }
```

Or like this if there are no arguments.
```
    public static @Nullable String getAlerts(@Nullable ThingActions actions) {
        if (actions == null) {
            throw new IllegalArgumentException("actions cannot be null");
        }
        Class<?>[] argTypes = {};
        return (String) invokeMethod(actions, "getAlerts", argTypes);
    }
```

Reply:
I hate to see parameter class types being defined for every single action method. 
My method would allow the code to be checked by the compiler for correctness.

```
    public static boolean acknowledge(@Nullable ThingActions actions, @Nullable String ackRef, @Nullable String ackType,
            @Nullable Boolean remindMeLater) {
        if (actions == null) {
            throw new IllegalArgumentException("actions cannot be null");
        }
        IEcobeeActions iactions = tryCast(actions);
        if(iactions != null){
            return iactions.acknowledge(ackRef, ackType, remindMeLater);
        }else{
            throw new IllegalArgumentException("actions is not a EcoBeeActions type");
        }
    }
```
If you wanted to you could throw these exceptions inside of `tryCast` to make the code even simpler:
```
    public static boolean acknowledge(@Nullable ThingActions actions, @Nullable String ackRef, @Nullable String ackType,
            @Nullable Boolean remindMeLater) {
        return tryCast(actions).acknowledge(ackRef, ackType, remindMeLater);
    }
```

I think proxies are probably the best workaround, even with the interface overhead which actually has the advantage of exposing the action api in a easy to read format that isn't obscured by annotations.
My method also has the ability to easily fall back to using the original ThingActions object directly in the case it is the correct one to begin with.

Reply:
Ah, now I see. Thanks for bearing with me...